akka-persistence/src/test/scala/akka/persistence/PersistentActorStashingSpec.scala

```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    val otherCommandHandler: Receive = {
+      case Cmd("c") ⇒
+        persist(Evt("c")) { event ⇒
+          updateState(event)
+          context.unbecome()
+        }
+        unstashAll()
+      case other ⇒ stash()
+    }
+  }
+
+  class AsyncStashingPersistentActor(name: String) extends ExamplePersistentActor(name) {
+    var stashed = false
+    val receiveCommand: Receive = commonBehavior orElse {
+      case Cmd("a") ⇒ persistAsync(Evt("a"))(updateState)
+      case Cmd("b") if !stashed ⇒
+        stash(); stashed = true
+      case Cmd("b") ⇒ persistAsync(Evt("b"))(updateState)
+      case Cmd("c") ⇒ persistAsync(Evt("c"))(updateState); unstashAll()
+    }
+  }
+
+}
+
+abstract class PersistentActorStashingSpec(config: Config) extends PersistenceSpec(config)
+  with ImplicitSender {
+
+  import PersistentActorStashingSpec._
+
+  "Stashing in a persistent actor" must {
+
+    "support user stash operations" in {
+      val persistentActor = namedPersistentActor[UserStashPersistentActor]
+      persistentActor ! Cmd("a")
+      persistentActor ! Cmd("b")
+      persistentActor ! Cmd("c")
+      expectMsg("b")
+      expectMsg("c")
+      expectMsg("a")
+    }
+
+    "support user stash operations with several stashed messages" in {
+      val persistentActor = namedPersistentActor[UserStashManyPersistentActor]
+      val n = 10
+      val cmds = 1 to n flatMap (_ ⇒ List(Cmd("a"), Cmd("b-1"), Cmd("b-2"), Cmd("c")))
+      val evts = 1 to n flatMap (_ ⇒ List("a", "c", "b-1", "b-2"))
+
+      cmds foreach (persistentActor ! _)
+      persistentActor ! GetState
+      expectMsg(evts)
+    }
+
+    "support user stash operations under failures" in {
+      val persistentActor = namedPersistentActor[UserStashFailurePersistentActor]
+      val bs = 1 to 10 map ("b-" + _)
+      persistentActor ! Cmd("a")
+      bs foreach (persistentActor ! Cmd(_))
+      persistentActor ! Cmd("c")
+      persistentActor ! GetState
+      expectMsg(List("a", "c") ++ bs.filter(_ != "b-2"))
+    }
+  }
+}
+
+class SteppingInMemPersistentActorStashingSpec extends PersistenceSpec(
+  SteppingInmemJournal.config("persistence-stash").withFallback(PersistenceSpec.config("stepping-inmem", "SteppingInMemPersistentActorStashingSpec")))
+  with ImplicitSender {
+
+  import PersistentActorStashingSpec._
+
+  "Stashing in a persistent actor mixed with persistAsync" should {
+
+    "handle async callback not happening until next message has been stashed" in {
+      val persistentActor = namedPersistentActor[AsyncStashingPersistentActor]
+      awaitAssert(SteppingInmemJournal.getRef("persistence-stash"), 3.seconds)
+      val journal = SteppingInmemJournal.getRef("persistence-stash")
+
+      // move the journal one step and wait for completion
+      def journalOperation(): Unit = {
+        journal ! SteppingInmemJournal.Token
+        expectMsg(SteppingInmemJournal.TokenConsumed)
+      }
```

Comment:
Yeah, since it's tests, and the in-mem journal we can just pick something. I'd say 3.seconds.dilated?